Graph apoptosis rollback re-adds node IDs as strings

## Context

Operator ran `/opt/homebrew/bin/npm run ai:run-sandman` from `/Users/Shared/github/neomjs/neo` on current `dev` (`dev == origin/dev`, `3271fb0281b2cd81fca84f403c0696b76e74d930`) on 2026-05-18.

The REM run reached graph garbage collection, detected `8408` orphaned nodes, then failed inside rollback:

```text
[INFO] [GraphMaintenanceService] Apoptosis detected 8408 orphaned nodes. Commencing eradication...
[ERROR] [DreamService] Failed to process undigested sessions: TypeError: Cannot create property 'Symbol(Neo.internalId)' on string 'CONCEPT:sunset-restart-cycle'
    at Store.assignInternalId (src/data/Store.mjs:411:30)
    at Store.splice (src/collection/Base.mjs:1530:33)
    at Store.splice (ai/graph/Store.mjs:161:30)
    at Store.add (src/collection/Base.mjs:181:21)
    at Store.add (src/data/Store.mjs:298:35)
    at Database.rollbackTransaction (ai/graph/Database.mjs:451:23)
    at Database.transaction (ai/graph/Database.mjs:482:18)
    at GraphService.removeNodes (ai/services/memory-core/GraphService.mjs:920:17)
    at GraphMaintenanceService.runGarbageCollection (ai/daemons/services/GraphMaintenanceService.mjs:53:26)
```

## The Problem

`GraphMaintenanceService.runGarbageCollection()` calls `GraphService.getOrphanedNodes()`, which returns an array of node ID strings. It then calls `GraphService.removeNodes(orphaned)`. `GraphService.removeNodes()` wraps `nodeIds.forEach(id => this.db.removeNode(id))` in `Database.transaction()`.

When the backing storage transaction fails, `Database.rollbackTransaction()` replays `transactionDiff` and does:

```js
if (mutation.removedItems?.length > 0) {
    store.add(mutation.removedItems);
}
```

The mutation buffer is not guaranteed to contain removed node objects. `src/collection/Base.mjs#splice()` returns actual removed objects, but its public `mutate` event emits:

```js
removedItems: toRemoveArray || removedItems
```

For remove-by-key calls, `toRemoveArray` is the string ID array. `Database.onNodesMutate()` stores that event payload in `transactionDiff`. Rollback then calls `store.add(['CONCEPT:sunset-restart-cycle', ...])`, and `Neo.data.Store#assignInternalId()` attempts to assign the internal ID symbol onto a string.

The failure is a rollback-path type-shape bug, not a data-specific issue with `CONCEPT:sunset-restart-cycle`.

## The Architectural Reality

- `ai/daemons/services/GraphMaintenanceService.mjs` owns the apoptosis pass and calls `GraphService.removeNodes(orphaned)`.
- `ai/services/memory-core/GraphService.mjs#getOrphanedNodes()` intentionally returns string IDs.
- `ai/graph/Database.mjs#transaction()` assumes rollback diff entries are re-addable item objects.
- `src/collection/Base.mjs#splice()` event payload uses remove input keys when remove-by-key was used, so transaction subscribers cannot assume `mutation.removedItems` contains objects.
- `ai/graph/storage/SQLite.mjs#removeNodes()` accepts either objects or strings, but rollback must restore in-memory stores, so it needs the actual removed node objects.

## The Fix

Choose the narrowest safe boundary after implementation V-B-A:

1. Normalize transaction diff capture in `Database.onNodesMutate()` / `onEdgesMutate()` so rollback receives actual removed objects, not remove keys; or
2. Change `Database.removeNode()` / `removeEdge()` transaction usage to capture the return value from `store.remove()` / `splice()` and push a rollback-safe diff; or
3. Change `Collection.splice()` mutate payload contract only if broader consumers expect actual removed objects. This is higher blast radius and should be avoided unless tests prove it is the correct substrate.

The likely safest local fix is inside `ai/graph/Database.mjs`: transaction rollback is the only consumer that requires object-shaped removed entries.

## Contract Ledger Matrix

| Target Surface | Source of Authority | Proposed Behavior | Fallback | Docs | Evidence |
|---|---|---|---|---|---|
| `Database.transaction()` rollback diff | This ticket + Sandman failure stack | Rollback re-adds object-shaped nodes/edges, never primitive remove keys | Abort rollback with explicit diagnostic before corrupting Store state | JSDoc on `rollbackTransaction()` | Unit test forces storage transaction failure after remove-by-key and verifies rollback completes |
| `GraphMaintenanceService.runGarbageCollection()` apoptosis | This ticket | Removing orphan string IDs either succeeds or fails without secondary rollback TypeError | Skip vector purge if graph removal failed | Existing Dream Pipeline docs | Re-run `ai:run-sandman` no longer throws `Cannot create property ... on string` |

## Acceptance Criteria

- [ ] Unit test reproduces a graph transaction failure after node removal by string ID and asserts rollback does not call `Store.add()` with strings.
- [ ] `GraphService.removeNodes(['CONCEPT:test'])` rollback preserves the original node object in memory when storage deletion fails.
- [ ] Edge rollback remains covered; removing edges by string ID must not regress.
- [ ] `GraphMaintenanceService.runGarbageCollection()` no longer throws `Cannot create property 'Symbol(Neo.internalId)' on string ...` when a bulk orphan deletion fails.
- [ ] Follow-up Sandman run reaches Golden Path synthesis without this rollback error.

## Out of Scope

- Changing apoptosis selection policy or protected node types.
- Deleting the 8408 orphan nodes manually.
- Reworking SQLite transaction semantics outside the rollback shape needed here.

## Avoided Traps

- **Treating `CONCEPT:sunset-restart-cycle` as corrupt data** — rejected. The stack proves the string type itself is the rollback payload shape.
- **Changing all collection mutation semantics globally first** — rejected until proven necessary; the graph transaction layer is the immediate consumer with the broken assumption.
- **Suppressing graph GC** — rejected. Apoptosis is expected maintenance; the rollback path must be safe.

## Related

- Operator Sandman run, 2026-05-18, current `dev` at `3271fb0281b2cd81fca84f403c0696b76e74d930`.
- `ai/daemons/services/GraphMaintenanceService.mjs:49-53`
- `ai/services/memory-core/GraphService.mjs:887-920`
- `ai/graph/Database.mjs:434-482`
- `src/collection/Base.mjs:1599-1614`

Origin Session ID: 8591bc48-0ddc-48bf-aa47-58e53ea81a57
Retrieval Hint: `query_raw_memories("Sandman GraphMaintenanceService apoptosis rollbackTransaction removedItems string CONCEPT:sunset-restart-cycle internalId")`

--- a/src/collection/Base.mjs
+++ b/src/collection/Base.mjs
@@ -1611,12 +1611,28 @@ class Collection extends Base {
                 // Always emit actual removed objects, not input keys. `removedItems` (local) is built
                 // at line 1487-1488 from `items.splice(indexOfKey(key), 1)[0]` — always object-shaped.
                 // The legacy `toRemoveArray || removedItems` fallback emitted the INPUT array, which
-                // contained STRING IDs when remove-by-key was used. Downstream consumers (SQLite.mjs
-                // line 279-287, Store.mjs line 164, Database.mjs onNodesMutate/onEdgesMutate +
-                // rollbackTransaction line 451) all expect OBJECT-shaped removedItems and silently
-                // failed (SQL no-ops) or loudly failed (rollback TypeError on string.Symbol-assignment)
-                // when fed string IDs. V-B-A 2026-05-18: only 2 mutate-event listeners exist in the
-                // codebase; both expect objects. Fix is empirically narrow-blast. (#11595)
+                // contained STRING IDs when remove-by-key was used. Rollback at Database.mjs:451
+                // (`store.add(mutation.removedItems)`) then attempted Symbol-assignment on primitive
+                // strings → TypeError (#11595 surface failure).
+                //
+                // Consumer-impact V-B-A 2026-05-18 (corrected scope per @neo-gpt PR #11611 Cycle 1
+                // review using `rg -n "mutate:|on(['\"]mutate" src ai`):
+                //   1. `Database.onNodesMutate` (ai/graph/Database.mjs:378) — consumes via
+                //      `storage.removeNodes` → SQLite.mjs:280 `node.id` extraction. REQUIRES object.
+                //   2. `Database.onEdgesMutate` (ai/graph/Database.mjs:358) — symmetric edge path.
+                //      REQUIRES object.
+                //   3. `Collection.Base.onMutate` (this file L1350) — source-bubble: calls
+                //      `me.splice(null, opts.removedItems, opts.addedItems)`. `splice` handles BOTH
+                //      shapes (L1483: `key = me.isItem(item) ? me.getKey(item) : item`). SHAPE-NEUTRAL.
+                //   4. `Data.Store.onCollectionMutate` (src/data/Store.mjs:1048) — uses
+                //      `opts.addedItems` only; does not touch `removedItems`. PAYLOAD-NEUTRAL.
+                //   5. `Grid.Container.onColumnsMutate` (src/grid/Container.mjs:997) — uses
+                //      `me._columns.items` directly; ignores mutation payload. PAYLOAD-NEUTRAL.
+                //
+                // Net: 2 consumers strictly require object-shape (and previously silently broke or
+                // loudly broke when fed strings); 3 are payload-compatible with either shape. Fix
+                // is structurally narrow-blast — no consumer breaks, 2 previously-silent bugs also
+                // fixed. (#11595)
                 removedItems   : removedItems
             })
         } else if (!me[silentUpdateMode]) {

--- a/src/collection/Base.mjs
+++ b/src/collection/Base.mjs
@@ -1608,7 +1608,16 @@ class Collection extends Base {
             me.fire('mutate', {
                 addedItems     : toAddArray,
                 preventBubbleUp: me.preventBubbleUp,
-                removedItems   : toRemoveArray || removedItems
+                // Always emit actual removed objects, not input keys. `removedItems` (local) is built
+                // at line 1487-1488 from `items.splice(indexOfKey(key), 1)[0]` — always object-shaped.
+                // The legacy `toRemoveArray || removedItems` fallback emitted the INPUT array, which
+                // contained STRING IDs when remove-by-key was used. Downstream consumers (SQLite.mjs
+                // line 279-287, Store.mjs line 164, Database.mjs onNodesMutate/onEdgesMutate +
+                // rollbackTransaction line 451) all expect OBJECT-shaped removedItems and silently
+                // failed (SQL no-ops) or loudly failed (rollback TypeError on string.Symbol-assignment)
+                // when fed string IDs. V-B-A 2026-05-18: only 2 mutate-event listeners exist in the
+                // codebase; both expect objects. Fix is empirically narrow-blast. (#11595)
+                removedItems   : removedItems
             })
         } else if (!me[silentUpdateMode]) {
             me.cacheUpdate({

--- a/test/playwright/unit/ai/graph/Database.spec.mjs
+++ b/test/playwright/unit/ai/graph/Database.spec.mjs
@@ -24,7 +24,7 @@ import path           from 'path';
 test.describe('Neo.ai.graph.Database', () => {
     let db;
     let testRun = 0;
-    
+
     // Build an isolated tmp path for the database file tests
     const tmpDir = path.resolve(process.cwd(), 'tmp');
     if (!fs.existsSync(tmpDir)) {
@@ -52,7 +52,7 @@ test.describe('Neo.ai.graph.Database', () => {
 
     test('should add and retrieve a node correctly', async () => {
         db.addNode({ id: 'node1', label: 'Person', properties: { name: 'Alice' } });
-        
+
         expect(db.nodes.getCount()).toBe(1);
         expect(db.nodes.get('node1').label).toBe('Person');
         expect(db.nodes.get('node1').properties.name).toBe('Alice');
@@ -61,9 +61,9 @@ test.describe('Neo.ai.graph.Database', () => {
     test('should add an edge correctly', async () => {
         db.addNode({ id: 'node1' });
         db.addNode({ id: 'node2' });
-        
+
         db.addEdge({ id: 'edge1', source: 'node1', target: 'node2', type: 'KNOWS' });
-        
+
         expect(db.edges.getCount()).toBe(1);
         expect(db.edges.get('edge1').type).toBe('KNOWS');
     });
@@ -72,10 +72,10 @@ test.describe('Neo.ai.graph.Database', () => {
         db.addNode({ id: 'node1' });
         db.addNode({ id: 'node2' });
         db.addNode({ id: 'node3' });
-        
+
         db.addEdge({ source: 'node1', target: 'node2', type: 'KNOWS' });
         db.addEdge({ source: 'node1', target: 'node3', type: 'LIKES' });
-        
+
         let adjacent = db.getAdjacentNodes('node1', 'outbound', 'KNOWS');
         expect(adjacent.length).toBe(1);
         expect(adjacent[0].id).toBe('node2');
@@ -87,12 +87,12 @@ test.describe('Neo.ai.graph.Database', () => {
     test('should cascade delete edges when a node is removed', async () => {
         db.addNode({ id: 'node1' });
         db.addNode({ id: 'node2' });
-        
+
         db.addEdge({ id: 'edge1', source: 'node1', target: 'node2', type: 'KNOWS' });
         expect(db.edges.getCount()).toBe(1);
 
         db.removeNode('node1');
-        
+
         expect(db.nodes.getCount()).toBe(1);
         expect(db.edges.getCount()).toBe(0); // Edge should be deleted because its source node is gone
     });
@@ -100,10 +100,10 @@ test.describe('Neo.ai.graph.Database', () => {
     test('should persist nodes and edges properly using SQLite storage adapter', async () => {
         // Clean out previous runs
         if (fs.existsSync(dbPath)) fs.unlinkSync(dbPath);
-        
+
         let storage = Neo.create(SQLite, { dbPath });
         await storage.initAsync();
-        
+
         let persistentDb = Neo.create(Database, {
             id: 'sqlite-graph-test',
             storage: storage
@@ -121,7 +121,7 @@ test.describe('Neo.ai.graph.Database', () => {
 
         let storageReload = Neo.create(SQLite, { dbPath });
         await storageReload.initAsync();
-        
+
         let reloadDb = Neo.create(Database, {
             id: 'sqlite-graph-reload',
             storage: storageReload
@@ -201,10 +201,10 @@ test.describe('Neo.ai.graph.Database', () => {
 
     test('should execute graph mutations synchronously within an atomic transaction', async () => {
         if (fs.existsSync(dbPath)) fs.unlinkSync(dbPath);
-        
+
         let storage = Neo.create(SQLite, { dbPath });
         await storage.initAsync();
-        
+
         let dbTransaction = Neo.create(Database, {
             id: 'sqlite-graph-transcation',
             storage: storage
@@ -214,21 +214,21 @@ test.describe('Neo.ai.graph.Database', () => {
             dbTransaction.addNode({ id: 'X' });
             dbTransaction.addNode({ id: 'Y' });
             dbTransaction.addEdge({ id: 'E1', source: 'X', target: 'Y', type: 'TEST' });
-            
-            // Nested checks should "see" data instantly in isolated synchronous thread natively 
+
+            // Nested checks should "see" data instantly in isolated synchronous thread natively
             expect(dbTransaction.nodes.getCount()).toBe(2);
             expect(dbTransaction.edges.getCount()).toBe(1);
         });
 
         // After transaction executes SQLite synchronously, verifying data survived cleanly natively
         expect(dbTransaction.nodes.getCount()).toBe(2);
-        
+
         // Ensure disk mappings captured transaction cleanly
         dbTransaction.destroy();
 
         let storageReload = Neo.create(SQLite, { dbPath });
         await storageReload.initAsync();
-        
+
         let reloadDb = Neo.create(Database, {
             id: 'sqlite-graph-txn-reload',
             storage: storageReload
@@ -236,7 +236,7 @@ test.describe('Neo.ai.graph.Database', () => {
         await storageReload.load();
 
         // Distributed Vicinity Load mapping Lazy Memory internally perfectly
-        reloadDb.getAdjacentNodes('X'); 
+        reloadDb.getAdjacentNodes('X');
 
         expect(reloadDb.edges.getCount()).toBe(1);
         reloadDb.destroy();
@@ -252,27 +252,98 @@ test.describe('Neo.ai.graph.Database', () => {
             dbRollback.transaction(() => {
                 dbRollback.addNode({ id: 'Poison' });
                 dbRollback.addEdge({ id: 'badEdge', source: 'Base', target: 'Poison' });
-                
-                // Assert it works momentarily 
+
+                // Assert it works momentarily
                 expect(dbRollback.nodes.getCount()).toBe(2);
-                
+
                 throw new Error("Triggered Exception!");
             });
         } catch(e) {
-            // Error intentionally caught gracefully natively 
+            // Error intentionally caught gracefully natively
         }
 
         // Memory rollback sequence has automatically fired instantly deleting uncommitted nodes cleanly!
         expect(dbRollback.nodes.getCount()).toBe(1);
         expect(dbRollback.edges.getCount()).toBe(0);
         expect(!dbRollback.nodes.has('Poison')).toBe(true);
-        
+
         dbRollback.destroy();
+    });
+
+    test('should rollback remove-by-key node removal without TypeError on string.Symbol assignment (#11595)', async () => {
+        // Regression #11595: GraphMaintenanceService apoptosis pass called
+        // GraphService.removeNodes(['CONCEPT:foo', ...]) inside a transaction; on rollback, the
+        // mutate event payload's removedItems contained STRING IDs (not actual node objects)
+        // because Collection.Base.splice() emitted `toRemoveArray || removedItems` — the INPUT
+        // key array, not the locally-computed actual-removed-objects array. Rollback then called
+        // store.add(stringIds) which triggered Store.assignInternalId() to attempt setting
+        // Symbol(Neo.internalId) on a primitive string → TypeError.
+        //
+        // Fix at src/collection/Base.mjs:1611: emit `removedItems` (local objects), not
+        // `toRemoveArray || removedItems`. This test exercises the exact failure path: remove
+        // node by key inside transaction → throw → rollback must restore the node-as-object
+        // without TypeError.
+        let dbRollbackByKey = Neo.create(Database, { id: 'graph-rollback-by-key-test' });
+
+        dbRollbackByKey.addNode({ id: 'CONCEPT:sunset-restart-cycle', type: 'CONCEPT', label: 'pre-remove-state' });
+        expect(dbRollbackByKey.nodes.getCount()).toBe(1);
+
+        try {
+            dbRollbackByKey.transaction(() => {
+                dbRollbackByKey.removeNode('CONCEPT:sunset-restart-cycle'); // remove by STRING key
+                expect(dbRollbackByKey.nodes.getCount()).toBe(0);
+                throw new Error('Simulated apoptosis pass failure (post-removal)');
+            });
+        } catch(e) {
+            expect(e.message).toBe('Simulated apoptosis pass failure (post-removal)');
+        }
+
+        // Rollback must restore the original node WITHOUT throwing TypeError on string-as-object.
+        expect(dbRollbackByKey.nodes.getCount()).toBe(1);
+        expect(dbRollbackByKey.nodes.has('CONCEPT:sunset-restart-cycle')).toBe(true);
+        // Verify the restored node is the original OBJECT (not a string wrapper).
+        const restored = dbRollbackByKey.nodes.get('CONCEPT:sunset-restart-cycle');
+        expect(typeof restored).toBe('object');
+        expect(restored.label).toBe('pre-remove-state');
+
+        dbRollbackByKey.destroy();
+    });
+
+    test('Collection.splice mutate-event removedItems is always object-shaped, regardless of remove-by-key vs remove-by-object input (#11595)', async () => {
+        // Direct contract test: the mutate event payload's removedItems must be the locally-built
+        // actual-removed-objects array, not the input keys. Empirical V-B-A on 2026-05-18: only 2
+        // mutate-event listeners exist in the codebase (Database.onEdgesMutate + onNodesMutate),
+        // both expect object-shaped payload. Fix at Collection.Base.splice() ensures consistency.
+        let dbContract = Neo.create(Database, { id: 'graph-mutate-contract-test' });
+        let capturedPayloads = [];
+
+        dbContract.addNode({ id: 'node-a', type: 'TEST' });
+        dbContract.nodes.on('mutate', mutation => {
+            if (mutation.removedItems?.length > 0) {
+                capturedPayloads.push(mutation.removedItems);
+            }
+        });
+
+        // Path 1: remove by STRING key
+        dbContract.nodes.remove('node-a');
+        expect(capturedPayloads.length).toBe(1);
+        expect(typeof capturedPayloads[0][0]).toBe('object');
+        expect(capturedPayloads[0][0].id).toBe('node-a');
+
+        // Path 2: remove by OBJECT
+        dbContract.addNode({ id: 'node-b', type: 'TEST' });
+        const nodeB = dbContract.nodes.get('node-b');
+        dbContract.nodes.remove(nodeB);
+        expect(capturedPayloads.length).toBe(2);
+        expect(typeof capturedPayloads[1][0]).toBe('object');
+        expect(capturedPayloads[1][0].id).toBe('node-b');
+
+        dbContract.destroy();
     });
 
     test('should enforce Cache Coherence flushing stale footprints dynamically directly mapping SQLite hardware triggers seamlessly', async () => {
         if (fs.existsSync(dbPath)) fs.unlinkSync(dbPath);
-        
+
         let storagePrimary = Neo.create(SQLite, { dbPath });
         await storagePrimary.initAsync();
         await storagePrimary.load();
@@ -291,10 +362,10 @@ test.describe('Neo.ai.graph.Database', () => {
         // Load Vicinity inside Secondary instance pulling it directly into Memory correctly
         dbSecondary.getAdjacentNodes('core');
         expect(dbSecondary.nodes.getCount()).toBe(2);
-        
+
         // Emulate Primary Server completely bypassing Secondary Process Memory internally modifying disk bounds directly!
-        dbPrimary.removeNode('branch'); 
-        
+        dbPrimary.removeNode('branch');
+
         // Automatically hardware Trigger executed locally recording branch deletion on GraphLog internally.
         // If Secondary Process hits an adjacent lookup, it automatically sweeps Garbage Logs!
         dbSecondary.getAdjacentNodes('core');
@@ -316,32 +387,32 @@ test.describe('Neo.ai.graph.Database', () => {
 
     test('should replay GraphLog mutations even on fresh boot when lastSyncId is 0 (Bug A)', async () => {
         if (fs.existsSync(dbPath)) fs.unlinkSync(dbPath);
-        
+
         let storageFresh = Neo.create(SQLite, { dbPath });
         await storageFresh.initAsync();
         let dbFresh = Neo.create(Database, { id: 'cache-fresh', storage: storageFresh });
         await storageFresh.load(); // DB empty -> lastSyncId = 0
-        
+
         expect(dbFresh.lastSyncId).toBe(0);
 
         let storageOther = Neo.create(SQLite, { dbPath });
         await storageOther.initAsync();
         let dbOther = Neo.create(Database, { id: 'cache-other', storage: storageOther });
         await storageOther.load();
-        
+
         dbOther.addNode({ id: 'race-node' });
-        
+
         // With Bug A fix, syncCache() processes the delta and lastSyncId advances
         dbFresh.syncCache();
         expect(dbFresh.lastSyncId).toBeGreaterThan(0);
-        
+
         dbFresh.destroy();
         dbOther.destroy();
     });
 
     test('should not mark vicinityLoadedNodes if lazy load returns empty (Bug B)', async () => {
         if (fs.existsSync(dbPath)) fs.unlinkSync(dbPath);
-        
+
         let storage = Neo.create(SQLite, { dbPath });
         await storage.initAsync();
         let db = Neo.create(Database, { id: 'bug-b', storage });
@@ -357,7 +428,7 @@ test.describe('Neo.ai.graph.Database', () => {
 
     test('should invalidate vicinity of both endpoints when syncCache invalidates edges (#10260)', async () => {
         if (fs.existsSync(dbPath)) fs.unlinkSync(dbPath);
-        
+
         // Setup Primary instance (Agent A)
         let storageA = Neo.create(SQLite, { dbPath });
         await storageA.initAsync();
@@ -378,7 +449,7 @@ test.describe('Neo.ai.graph.Database', () => {
         dbB.syncCache();
         dbB.getAdjacentNodes('node-x');
         dbB.getAdjacentNodes('node-y');
-        
+
         expect(dbB.vicinityLoadedNodes.has('node-x')).toBe(true);
         expect(dbB.vicinityLoadedNodes.has('node-y')).toBe(true);
         expect(dbB.edges.getCount()).toBe(0);
@@ -386,10 +457,10 @@ test.describe('Neo.ai.graph.Database', () => {
         // Instance A adds a new edge between the nodes
         dbA.addEdge({ id: 'edge-xy', source: 'node-x', target: 'node-y', type: 'LINKS' });
 
-        // Without the #10260 fix, dbB.syncCache() would remove the edge ID from delta, 
+        // Without the #10260 fix, dbB.syncCache() would remove the edge ID from delta,
         // but it wouldn't clear 'node-x' or 'node-y' from vicinityLoadedNodes.
         dbB.syncCache();
-        
+
         // With #10260, they should no longer be marked as loaded
         expect(dbB.vicinityLoadedNodes.has('node-x')).toBe(false);
         expect(dbB.vicinityLoadedNodes.has('node-y')).toBe(false);

--- a/test/playwright/unit/ai/graph/Database.spec.mjs
+++ b/test/playwright/unit/ai/graph/Database.spec.mjs
@@ -309,11 +309,50 @@ test.describe('Neo.ai.graph.Database', () => {
         dbRollbackByKey.destroy();
     });
 
+    test('should rollback remove-by-key edge removal without TypeError (#11595)', async () => {
+        // Sibling regression #11595: same shape-mismatch class on the edge-removal path.
+        // `Database.removeEdge(edgeId)` flows STRING IDs through the mutate event. Rollback must
+        // restore the edge as an OBJECT without TypeError on string.Symbol assignment. Per #11595
+        // AC explicit requirement (caught by @neo-gpt PR #11611 Cycle 1 review): "Edge rollback
+        // remains covered; removing edges by string ID must not regress."
+        let dbEdgeRollback = Neo.create(Database, { id: 'graph-edge-rollback-test' });
+
+        dbEdgeRollback.addNode({ id: 'A', label: 'src' });
+        dbEdgeRollback.addNode({ id: 'B', label: 'dst' });
+        dbEdgeRollback.addEdge({ id: 'AB', source: 'A', target: 'B', type: 'TEST' });
+        expect(dbEdgeRollback.edges.getCount()).toBe(1);
+        expect(dbEdgeRollback.edges.get('AB').type).toBe('TEST');
+
+        try {
+            dbEdgeRollback.transaction(() => {
+                dbEdgeRollback.removeEdge('AB'); // remove by STRING key
+                expect(dbEdgeRollback.edges.getCount()).toBe(0);
+                throw new Error('Simulated post-edge-removal failure');
+            });
+        } catch(e) {
+            expect(e.message).toBe('Simulated post-edge-removal failure');
+        }
+
+        // Rollback must restore the original edge OBJECT (not a string wrapper).
+        expect(dbEdgeRollback.edges.getCount()).toBe(1);
+        expect(dbEdgeRollback.edges.has('AB')).toBe(true);
+        const restoredEdge = dbEdgeRollback.edges.get('AB');
+        expect(typeof restoredEdge).toBe('object');
+        expect(restoredEdge.source).toBe('A');
+        expect(restoredEdge.target).toBe('B');
+        expect(restoredEdge.type).toBe('TEST');
+
+        dbEdgeRollback.destroy();
+    });
+
     test('Collection.splice mutate-event removedItems is always object-shaped, regardless of remove-by-key vs remove-by-object input (#11595)', async () => {
         // Direct contract test: the mutate event payload's removedItems must be the locally-built
-        // actual-removed-objects array, not the input keys. Empirical V-B-A on 2026-05-18: only 2
-        // mutate-event listeners exist in the codebase (Database.onEdgesMutate + onNodesMutate),
-        // both expect object-shaped payload. Fix at Collection.Base.splice() ensures consistency.
+        // actual-removed-objects array, not the input keys. V-B-A on consumer impact across the
+        // 5 mutate-event listeners in the codebase (per @neo-gpt PR #11611 Cycle 1 review):
+        // 2 require object-shape (Database.onEdgesMutate + onNodesMutate); 3 are payload-neutral
+        // or shape-flexible (Collection.Base.onMutate forwards via splice which handles both,
+        // Data.Store.onCollectionMutate uses only addedItems, Grid.Container.onColumnsMutate
+        // ignores mutation payload). Fix at Collection.Base.splice() is consumer-safe.
         let dbContract = Neo.create(Database, { id: 'graph-mutate-contract-test' });
         let capturedPayloads = [];
 

